hashicorp/consul-k8s
# [NET-10567] Fix namespace normalization on external registration/ACL Setup for Terminating Gateways

### Changes proposed in this PR ###  
- Fix namespace normalization on external registration (when namespaces are enabled the empty namespace is converted to "default" for acl rules)
- Fix how acls are setup for terminating gateways -> moved from the registrations controller/cache to the terminating gateway controller to avoid race conditions with persisting the terminating gateway config entry (TODO: clean up extra policies when a terminating gateway is removed/services are removed and are not referenced by any other terminating gateways)
- Fixed bug in cache where values weren't being persisted with the correct key
- updated terminating gateway tests so they would catch this in the future

### How I've tested this PR ###
- wrote tests
- ran in kind cluster

### How I expect reviewers to test this PR ###
- clone this branch and build the container using `make docker-dev`
- clone down the following gist: https://gist.github.com/jm96441n/1be432dd2b6b5ade7bc1b058539f6095
- modify the `CONSUL_K8S_CHART_LOCATION` variable in the `start.sh` file to point to the helm charts in your local version of consul-
- run the `start.sh` file (this requires you to have `kind` and `yq` on your machine, and you'll need to run `chmod +x ./start.sh`)
- after that completes run the following to set up your consul CLI to talk to consul:
```sh
export CONSUL_HTTP_ADDR="127.0.0.1:8501"
export CONSUL_HTTP_TOKEN=$(kubectl get --namespace consul secrets/consul-consul-bootstrap-acl-token --template={{.data.token}} | base64 -d)
export CONSUL_HTTP_SSL=true
export CONSUL_HTTP_SSL_VERIFY=false
```
- enable a port forward to the consul server on port 8501
- run `consul acl role list` and you will see both terminating gateways with the term gateway policy and the zoidberg and nibbler policies
- run `consul acl policy read -name zoidberg-write-policy` to see the policy include the namespace
- shell into the bender service and run `curl localhost:1234` to see the request to zoidberg go through, run `curl localhost:5678` to see the request to nibbler go through
- now we'll test that acl policies are removed correctly
- edit the `termgw.yaml` file to no longer reference the zoidberg service and apply the file
- run `consul acl role list` and see the first terminating gateway no longer references the zoidberg policy
- run `consul acl policy list` and you'll see the zoidberg policy still exists
- now edit the termgw2.yaml file to no longer reference the zoidberg service and apply that file
- now run `consul acl role list` and see that none of the terminating gateways reference that policy
- now run `consul acl policy list` and see that the zoidberg policy no longer exists because no gateway is referencing it


### Checklist ###
